### PR TITLE
[openhab/jamod] Remove org.apache.commons.* dependencies

### PR DESCRIPTION
**Pull request by andrewfg**
- Remove org.apache.commons.io.* dependencies
- Remove org.apache.commons.lang.* dependencies

References..
- https://github.com/openhab/openhab-core/issues/2115
- https://github.com/openhab/openhab-core/pull/2127

@ssalonen @splatch @cweitkamp

NOTES
1. I am not 100% comfortable with my 'brute' refactoring of IOUtils.closeQuietly() => please check!!
2. The PR https://github.com/openhab/openhab-core/pull/2127 depends on a new version of this JAR file, so merges must be done in the right order!

### CODE REVIEW

File: src/main/java/net/wimpi/modbus/net/TCPMasterConnection.java [hunk 1]
```diff
@@ -233,7 +225,9 @@ public void resetConnection() {
 
     @Override
     public String toString() {
-        return new ToStringBuilder(this, toStringStyle).append("socket", m_Socket).toString();
+        return "TCPMasterConnection [m_Socket=" + m_Socket + ", m_Timeout=" + m_Timeout + ", m_Connected=" + m_Connected
```
Comment (ssalonen) [inline src/main/java/net/wimpi/modbus/net/TCPMasterConnection.java:228]: Risk of NPE? E.g. Socket can be null

File: src/main/java/net/wimpi/modbus/net/SerialConnection.java [hunk 2]
```diff
@@ -338,8 +330,7 @@ public boolean isConnected() {
 
     @Override
     public String toString() {
-        return new ToStringBuilder(this, toStringStyle).append("portName", m_Parameters.getPortName())
-                .append("port", m_SerialPort).toString();
+        return "SerialConnection [m_SerialPort=" + m_SerialPort + ", m_Parameters.getPortName()="
```
Comment (ssalonen) [inline src/main/java/net/wimpi/modbus/net/SerialConnection.java]: Risk of NPE

File: src/main/java/net/wimpi/modbus/net/UDPMasterConnection.java [hunk 3]
```diff
@@ -195,7 +186,7 @@ public void resetConnection() {
 
     @Override
     public String toString() {
-        return new ToStringBuilder(this, toStringStyle).append("terminal", m_Terminal).toString();
+        return "UDPMasterConnection [m_Terminal=" + m_Terminal + ", m_Timeout=" + m_Timeout + ", m_Connected="
```
Comment (ssalonen) [inline src/main/java/net/wimpi/modbus/net/UDPMasterConnection.java:189]: Risk of NPE

File: src/main/java/net/wimpi/modbus/util/SerialParameters.java [hunk 4]
```diff
@@ -97,18 +87,16 @@ public SerialParameters(String portName, int baudRate, int flowControlIn, int fl
 
     @Override
     public String toString() {
-        return new ToStringBuilder(this, toStringStyle).append("portName", m_PortName).append("baudRate", m_BaudRate)
-                .append("flowControlIn", getFlowControlInString()).append("flowControlOut", getFlowControlOutString())
-                .append("databits", getDatabitsString()).append("stopbits", getStopbitsString())
-                .append("parity", getParityString()).append("encoding", m_Encoding).append("echo", m_Echo)
-                .append("receiveTimeoutMillis", m_ReceiveTimeoutMillis).toString();
+        return "SerialParameters [m_PortName=" + m_PortName + ", m_BaudRate=" + m_BaudRate + ", m_FlowControlIn="
```
Comment (ssalonen) [inline src/main/java/net/wimpi/modbus/util/SerialParameters.java:90]: Should we prepare for null values? Not perhaps very likely in this specific method

**Review by ssalonen — CHANGES_REQUESTED** [review]

**Comment by ssalonen** [comment]
Thanks @andrewfg .

To me the closequietly replacement is correctly implementated

File: src/main/java/net/wimpi/modbus/net/TCPMasterConnection.java (hunk 1 above)
Comment (andrewfg) [reply src/main/java/net/wimpi/modbus/net/TCPMasterConnection.java:228]: Yes, I suppose you are right.

File: src/main/java/net/wimpi/modbus/net/SerialConnection.java (hunk 2 above)
Comment (andrewfg) [reply src/main/java/net/wimpi/modbus/net/SerialConnection.java]: Ditto. I used the Eclipse macro to create these methods, but it seems not to be aware of null checks. Odd.

File: src/main/java/net/wimpi/modbus/net/UDPMasterConnection.java (hunk 3 above)
Comment (andrewfg) [reply src/main/java/net/wimpi/modbus/net/UDPMasterConnection.java:189]: Yup.

File: src/main/java/net/wimpi/modbus/util/SerialParameters.java (hunk 4 above)
Comment (andrewfg) [reply src/main/java/net/wimpi/modbus/util/SerialParameters.java:90]: Let me think about this..

**Comment by cweitkamp** [comment]
You maybe directly want to remove the dependencies from the POM:

https://github.com/openhab/jamod/blob/b2c2692c8ea69e93070eb24b06900aba75f5ff3e/pom.xml#L66-L77

This additionally ensures the code will compile without apache.commons.

File: src/main/java/net/wimpi/modbus/net/TCPMasterConnection.java (hunk 1 above)
Comment (andrewfg) [reply src/main/java/net/wimpi/modbus/net/TCPMasterConnection.java:228]: No. Actually I was right after all. 

The function of Java string concatenation (like `z= a + b + c`) is really clever; it does the following things implicitly:
1. If an argument is a class other than String then Java expands `argument` to `argument.toString()`
2. If an argument is `null` then Java inserts "null" in the resulting string; and it does not create an NPE

Try the following code..
```
    @Test
    void test() {
        String a = "the" + "quick" + null + "brown";
        assertEquals(a, "thequicknullbrown");
        System.out.println(a);
        System.out.println();

        String b = "the" + "quick" + this + "brown";
        assertTrue(b.startsWith("thequick"));
        System.out.println(b);
        System.out.println();

        Object c;
        for (int i = 0; i < 5; i++) {
            if (i % 2 == 0) {
                c = null;
            } else {
                c = "    ";
            }
            String d = "the" + "quick" + c + "brown";
            assertTrue(b.startsWith("thequick"));
            System.out.println(d);
        }
    }
```

File: src/main/java/net/wimpi/modbus/net/SerialConnection.java (hunk 2 above)
Comment (andrewfg) [reply src/main/java/net/wimpi/modbus/net/SerialConnection.java:333]: See above..

Comment (andrewfg) [reply src/main/java/net/wimpi/modbus/net/SerialConnection.java]: However, here there is indeed a potential for NPE

File: src/main/java/net/wimpi/modbus/net/UDPMasterConnection.java (hunk 3 above)
Comment (andrewfg) [reply src/main/java/net/wimpi/modbus/net/UDPMasterConnection.java:189]: Nope. See above..

File: src/main/java/net/wimpi/modbus/util/SerialParameters.java (hunk 4 above)
Comment (andrewfg) [reply src/main/java/net/wimpi/modbus/util/SerialParameters.java:90]: See above..

File: src/main/java/net/wimpi/modbus/net/SerialConnection.java (hunk 2 above)
Comment (andrewfg) [reply src/main/java/net/wimpi/modbus/net/SerialConnection.java]: Fixed

File: src/main/java/net/wimpi/modbus/net/TCPMasterConnection.java (hunk 1 above)
Comment (ssalonen) [reply src/main/java/net/wimpi/modbus/net/TCPMasterConnection.java:228]: Thanks, I learned something new 🤯

Comment (andrewfg) [reply src/main/java/net/wimpi/modbus/net/TCPMasterConnection.java:228]: Me too. Your comment was the trigger for me to look it up.

**Comment by splatch** [comment]
:+1: LGTM

**Comment by kaikreuzer** [comment]
@ssalonen Does this get your approval?

**Review by ssalonen — APPROVED** [review]

**Comment by ssalonen** [comment]
Absolutely! Good work.

**Comment by andrewfg** [comment]
NB the removal of the Apache dependencies does not actually change the interface between jamod and the OH core module. So I think the solution is just to increment the version number in the pom of this jamod PR, and in the OH core allow both the prior and the new versions of jamod in the core module’s respective pom dependency entry. Or??

**Comment by andrewfg** [comment]
Bumped version ID

**Comment by andrewfg** [comment]
PS you might want to merge https://github.com/openhab/jamod/pull/7 as well (???)

**Comment by kaikreuzer** [comment]
New version has been published: https://bintray.com/openhab/mvn/jamod/1.3.1.OH

**Comment by andrewfg** [comment]
Thank you @kaikreuzer

**Pull request merged**